File: src/services/datasetService.ts
/**
 * Nexora Dataset Service
 * Provides real-time access to AI4I 2020 Predictive Maintenance and
 * Smoke/Fire IoT datasets, plus synthesized PPE detection events.
 *
 * Supports two modes:
 *   DATASET_REPLAY  — cycles through real dataset rows at configurable speed
 *   SIMULATION      — scenario-driven event generation
 */

import { AI4I_DATASET, AI4IRecord } from '../data/ai4iData';
import { SMOKE_DATASET, SmokeRecord } from '../data/smokeData';

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

export type DataMode = 'SIMULATION' | 'DATASET_REPLAY';

export type PPEViolationType = 'HELMET_MISSING' | 'VEST_MISSING' | 'GLOVES_MISSING' | 'GOGGLES_MISSING' | 'ALL_COMPLIANT';

export type VisionDetectionType =
  | 'PPE_VIOLATION'
  | 'FIRE_DETECTED'
  | 'SMOKE_DETECTED'
  | 'RESTRICTED_ZONE_ENTRY'
  | 'WORKER_DOWN'
  | 'ALL_CLEAR';

export interface VisionDetection {
  id: string;
  cameraId: string;
  cameraZone: string;
  detectionType: VisionDetectionType;
  subType?: PPEViolationType;
  confidence: number;         // 0–100
  timestamp: string;
  riskBoost: number;          // Points added to compound risk
  workerName?: string;
  imageTag: string;           // Descriptive label for Evidence Viewer
  notified: boolean;
  severity: 'CRITICAL' | 'HIGH' | 'WARNING' | 'SAFE';
}

export interface MachineHealthSnapshot {
  assetId: string;
  assetName: string;
  productId: string;
  airTempC: number;
  processTempC: number;
  rpm: number;
  torqueNm: number;
  toolWearMin: number;
  healthScore: number;            // 0–100
  failureProbability: number;     // 0–100
  remainingUsefulLife: number;    // hours
  failureMode: string | null;
  machineFailure: boolean;
  dataSource: 'AI4I_DATASET' | 'SIMULATION';
}

export interface SmokeFireSnapshot {
  tempC: number;
  humidity: number;
  tvocPpb: number;
  eco2Ppm: number;
  pm25: number;
  fireAlarm: boolean;
  smokeLevel: 'NONE' | 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL';
  dataSource: 'SMOKE_DATASET' | 'SIMULATION';
}

// ─────────────────────────────────────────────────────────────────────────────
// Camera Configuration
// ─────────────────────────────────────────────────────────────────────────────

export const CAMERA_CONFIG = [
  { id: 'CAM-C01', zone: 'Zone A1', name: 'Distillation Area — West Entry', position: { x: 110, y: 80 } },
  { id: 'CAM-C02', zone: 'Zone A1', name: 'Pump P-201 Area',               position: { x: 150, y: 330 } },
  { id: 'CAM-C03', zone: 'Zone B4', name: 'Reactor Feed — Hot Work Zone',  position: { x: 445, y: 80 } },
  { id: 'CAM-C04', zone: 'Zone B4', name: 'Pump P-102 Critical Zone',      position: { x: 445, y: 340 } },
  { id: 'CAM-C05', zone: 'Zone B4', name: 'Gas Sensor AT-G104 Proximity',  position: { x: 560, y: 130 } },
  { id: 'CAM-C06', zone: 'Zone C2', name: 'Separator V-401 Area',          position: { x: 790, y: 100 } },
] as const;

export type CameraId = typeof CAMERA_CONFIG[number]['id'];

// ─────────────────────────────────────────────────────────────────────────────
// Risk contribution map
// ─────────────────────────────────────────────────────────────────────────────

export const VISION_RISK_BOOSTS: Record<VisionDetectionType, number> = {
  FIRE_DETECTED:        40,
  SMOKE_DETECTED:       25,
  RESTRICTED_ZONE_ENTRY: 18,
  PPE_VIOLATION:        10,
  WORKER_DOWN:          30,
  ALL_CLEAR:            0,
};

// ─────────────────────────────────────────────────────────────────────────────
// Predictive Maintenance — AI4I Dataset Service
// ─────────────────────────────────────────────────────────────────────────────

/** Derive machine health snapshot from a real AI4I record */
export function getMachineSnapshot(record: AI4IRecord, assetId = 'PUMP-P102'): MachineHealthSnapshot {
  const processTempC = record.processTempK - 273.15;
  const airTempC = record.airTempK - 273.15;

  // Health degrades with tool wear and torque deviation from nominal (40 Nm)
  const torqueDeviation = Math.abs(record.torqueNm - 40) / 40;
  const wearFactor = Math.min(1, record.toolWearMin / 200);
  const healthScore = Math.max(5, Math.round(100 - (wearFactor * 50) - (torqueDeviation * 30) - (record.machineFailure * 35)));

  // Failure probability: logistic-style curve from health
  const failureProbability = record.machineFailure === 1
    ? Math.round(75 + Math.random() * 20)
    : Math.round(Math.max(2, (1 - healthScore / 100) * 60));

  // RUL: remaining useful life (hours) — inverse of wear progression
  const remainingUsefulLife = record.machineFailure === 1
    ? Math.round(Math.random() * 12)
    : Math.round((200 - record.toolWearMin) * 0.8);

  // Determine active failure mode
  let failureMode: string | null = null;
  if (record.twf) failureMode = 'Tool Wear Failure (TWF)';
  else if (record.hdf) failureMode = 'Heat Dissipation Failure (HDF)';
  else if (record.pwf) failureMode = 'Power Failure (PWF)';
  else if (record.osf) failureMode = 'Overstrain Failure (OSF)';
  else if (record.rnf) failureMode = 'Random Failure (RNF)';

  return {
    assetId,
    assetName: assetId === 'PUMP-P102' ? 'Hydrocracker Charge Pump P-102' : `Machine ${record.productId}`,
    productId: record.productId,
    airTempC: +airTempC.toFixed(1),
    processTempC: +processTempC.toFixed(1),
    rpm: record.rpm,
    torqueNm: record.torqueNm,
    toolWearMin: record.toolWearMin,
    healthScore,
    failureProbability,
    remainingUsefulLife,
    failureMode,
    machineFailure: record.machineFailure === 1,
    dataSource: 'AI4I_DATASET',
  };
}

/** Get AI4I record at playback index (wraps around) */
export function getAI4IRecord(index: number): AI4IRecord {
  return AI4I_DATASET[index % AI4I_DATASET.length];
}

/** Get total AI4I dataset length */
export const AI4I_LENGTH = AI4I_DATASET.length;

// ─────────────────────────────────────────────────────────────────────────────
// Smoke / Fire — IoT Dataset Service
// ─────────────────────────────────────────────────────────────────────────────

export function getSmokeSnapshot(record: SmokeRecord): SmokeFireSnapshot {
  let smokeLevel: SmokeFireSnapshot['smokeLevel'] = 'NONE';
  if (record.fireAlarm) {
    if (record.tvoc > 5000 || record.pm25 > 10) smokeLevel = 'CRITICAL';
    else if (record.tvoc > 1000 || record.pm25 > 5) smokeLevel = 'HIGH';
    else smokeLevel = 'MODERATE';
  } else if (record.tvoc > 100 || record.pm25 > 2) {
    smokeLevel = 'LOW';
  }

  return {
    tempC: +record.tempC.toFixed(1),
    humidity: +record.humidity.toFixed(1),
    tvocPpb: +record.tvoc.toFixed(0),
    eco2Ppm: +record.eco2.toFixed(0),
    pm25: +record.pm25.toFixed(2),
    fireAlarm: record.fireAlarm === 1,
    smokeLevel,
    dataSource: 'SMOKE_DATASET',
  };
}

export function getSmokeRecord(index: number): SmokeRecord {
  return SMOKE_DATASET[index % SMOKE_DATASET.length];
}

export const SMOKE_LENGTH = SMOKE_DATASET.length;

// ─────────────────────────────────────────────────────────────────────────────
// PPE Detection — Synthesized Event Stream
// ─────────────────────────────────────────────────────────────────────────────

const PPE_WORKER_POOL = ['W-804 (Vikram Sharma)', 'W-809 (Arjun Mehta)', 'W-302 (Priya Nair)', 'W-115 (Rajesh Kumar)', 'W-712 (Deepak Rao)'];

const PPE_EVENT_POOL: Array<Omit<VisionDetection, 'id' | 'timestamp'>> = [
  { cameraId: 'CAM-C03', cameraZone: 'Zone B4', detectionType: 'PPE_VIOLATION', subType: 'HELMET_MISSING',
    confidence: 98, riskBoost: 10, workerName: PPE_WORKER_POOL[0], imageTag: 'Worker without helmet near hot work zone', notified: true, severity: 'HIGH' },
  { cameraId: 'CAM-C04', cameraZone: 'Zone B4', detectionType: 'PPE_VIOLATION', subType: 'VEST_MISSING',
    confidence: 94, riskBoost: 10, workerName: PPE_WORKER_POOL[1], imageTag: 'Safety vest non-compliance detected', notified: false, severity: 'WARNING' },
  { cameraId: 'CAM-C05', cameraZone: 'Zone B4', detectionType: 'RESTRICTED_ZONE_ENTRY',
    confidence: 97, riskBoost: 18, workerName: PPE_WORKER_POOL[2], imageTag: 'Unauthorized entry into gas hazard zone', notified: true, severity: 'CRITICAL' },
  { cameraId: 'CAM-C01', cameraZone: 'Zone A1', detectionType: 'PPE_VIOLATION', subType: 'GLOVES_MISSING',
    confidence: 89, riskBoost: 10, workerName: PPE_WORKER_POOL[3], imageTag: 'Chemical gloves missing during equipment handling', notified: false, severity: 'WARNING' },
  { cameraId: 'CAM-C06', cameraZone: 'Zone C2', detectionType: 'PPE_VIOLATION', subType: 'GOGGLES_MISSING',
    confidence: 92, riskBoost: 10, workerName: PPE_WORKER_POOL[4], imageTag: 'Safety goggles not worn near separator vessel', notified: true, severity: 'WARNING' },
  { cameraId: 'CAM-C04', cameraZone: 'Zone B4', detectionType: 'RESTRICTED_ZONE_ENTRY',
    confidence: 99, riskBoost: 18, workerName: PPE_WORKER_POOL[0], imageTag: 'Worker within 3m of CRITICAL gas plume boundary', notified: true, severity: 'CRITICAL' },
];

let _ppeEventIndex = 0;

export function getNextPPEEvent(): VisionDetection {
  const base = PPE_EVENT_POOL[_ppeEventIndex % PPE_EVENT_POOL.length];
  _ppeEventIndex++;
  return {
    ...base,
    id: `VIS-${Date.now()}-${_ppeEventIndex}`,
    timestamp: new Date().toISOString(),
  };
}

/** Generate fire/smoke detection event from a real smoke dataset record */
export function getSmokeVisionEvent(record: SmokeRecord, cameraId: CameraId = 'CAM-C03'): VisionDetection | null {
  const snapshot = getSmokeSnapshot(record);
  const cam = CAMERA_CONFIG.find(c => c.id === cameraId)!;

  if (snapshot.smokeLevel === 'NONE') return null;

  const isFire = snapshot.fireAlarm && snapshot.smokeLevel === 'CRITICAL';
  const detectionType: VisionDetectionType = isFire ? 'FIRE_DETECTED' : 'SMOKE_DETECTED';

  return {
    id: `VIS-${Date.now()}-SMOKE`,
    cameraId,
    cameraZone: cam.zone,
    detectionType,
    confidence: isFire ? 97 : 88,
    timestamp: new Date().toISOString(),
    riskBoost: VISION_RISK_BOOSTS[detectionType],
    imageTag: isFire
      ? `Fire detected — Temp ${snapshot.tempC}°C, TVOC ${snapshot.tvocPpb} ppb, PM2.5 ${snapshot.pm25}`
      : `Smoke detected — TVOC ${snapshot.tvocPpb} ppb, PM2.5 ${snapshot.pm25}, eCO2 ${snapshot.eco2Ppm} ppm`,
    notified: true,
    severity: isFire ? 'CRITICAL' : 'HIGH',
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// Simulation Scenarios
// ─────────────────────────────────────────────────────────────────────────────

export type SimulationScenario =
  | 'GAS_LEAK'
  | 'FIRE'
  | 'SMOKE'
  | 'PPE_VIOLATION'
  | 'MACHINE_FAILURE'
  | 'MULTI_INCIDENT'
  | 'NONE';

export interface ScenarioConfig {
  id: SimulationScenario;
  label: string;
  description: string;
  riskBoost: number;
  affectedZones: string[];
  affectedCameras: CameraId[];
  icon: string;
  severity: 'CRITICAL' | 'HIGH' | 'WARNING';
}

export const SIMULATION_SCENARIOS: ScenarioConfig[] = [
  {
    id: 'GAS_LEAK',
    label: 'Gas Leak Event',
    description: 'Hydrocarbon vapor accumulation in Zone B4 exceeds 20% LEL',
    riskBoost: 35,
    affectedZones: ['Zone B4'],
    affectedCameras: ['CAM-C03', 'CAM-C04', 'CAM-C05'],
    icon: '💨',
    severity: 'CRITICAL',
  },
  {
    id: 'FIRE',
    label: 'Fire Detection',
    description: 'Open flame detected at Zone B4 by thermal camera network',
    riskBoost: 40,
    affectedZones: ['Zone B4'],
    affectedCameras: ['CAM-C03', 'CAM-C04'],
    icon: '🔥',
    severity: 'CRITICAL',
  },
  {
    id: 'SMOKE',
    label: 'Smoke Detection',
    description: 'IoT smoke sensors detect TVOC spike — possible smoldering event',
    riskBoost: 25,
    affectedZones: ['Zone B4', 'Zone A1'],
    affectedCameras: ['CAM-C03', 'CAM-C01'],
    icon: '🌫️',
    severity: 'HIGH',
  },
  {
    id: 'PPE_VIOLATION',
    label: 'PPE Violation',
    description: 'AI Vision detects worker without compliant personal protective equipment',
    riskBoost: 10,
    affectedZones: ['Zone B4'],
    affectedCameras: ['CAM-C04'],
    icon: '🦺',
    severity: 'WARNING',
  },
  {
    id: 'MACHINE_FAILURE',
    label: 'Machine Failure',
    description: 'AI4I dataset predicts imminent bearing failure on Pump P-102',
    riskBoost: 20,
    affectedZones: ['Zone B4'],
    affectedCameras: ['CAM-C04'],
    icon: '⚙️',
    severity: 'CRITICAL',
  },
  {
    id: 'MULTI_INCIDENT',
    label: 'Multi-Incident',
    description: 'Concurrent: Gas Leak + PPE Violation + Machine Failure in Zone B4',
    riskBoost: 55,
    affectedZones: ['Zone B4', 'Zone A1', 'Zone C2'],
    affectedCameras: ['CAM-C03', 'CAM-C04', 'CAM-C05', 'CAM-C06'],
    icon: '🚨',
    severity: 'CRITICAL',
  },
];

File: backend/services/dataset_service.py
"""
Backend Dataset Service
Parses AI4I 2020 Predictive Maintenance & Smoke/Fire IoT datasets directly from CSV files.
Provides real-time machine health records, smoke/fire data streams, and synthesized PPE events.
"""

import os
import csv
import math
from typing import List, Dict, Any, Optional

class DatasetService:
    def __init__(self):
        self.ai4i_records: List[Dict[str, Any]] = []
        self.smoke_records: List[Dict[str, Any]] = []
        self.ai4i_index = 0
        self.smoke_index = 0
        
        self._load_ai4i_dataset()
        self._load_smoke_dataset()

    def _load_ai4i_dataset(self):
        path = os.path.join("Datasets", "ai4i+2020+predictive+maintenance+dataset", "ai4i2020.csv")
        if not os.path.exists(path):
            print(f"[DatasetService] Warning: {path} not found.")
            return

        with open(path, mode="r", encoding="utf-8-sig") as f:
            reader = csv.DictReader(f)
            for row in reader:
                try:
                    udi_key = [k for k in row.keys() if "UDI" in k][0]
                    self.ai4i_records.append({
                        "udi": int(row[udi_key]),
                        "product_id": row["Product ID"],
                        "type": row["Type"],
                        "air_temp_k": float(row["Air temperature [K]"]),
                        "process_temp_k": float(row["Process temperature [K]"]),
                        "rpm": int(float(row["Rotational speed [rpm]"])),
                        "torque_nm": float(row["Torque [Nm]"]),
                        "tool_wear_min": int(float(row["Tool wear [min]"])),
                        "machine_failure": int(row["Machine failure"]),
                        "twf": int(row["TWF"]),
                        "hdf": int(row["HDF"]),
                        "pwf": int(row["PWF"]),
                        "osf": int(row["OSF"]),
                        "rnf": int(row["RNF"]),
                    })
                except Exception:
                    continue

        print(f"[DatasetService] Loaded {len(self.ai4i_records)} AI4I records.")

    def _load_smoke_dataset(self):
        path = os.path.join("Datasets", "archive (3)", "smoke_detection_iot.csv")
        if not os.path.exists(path):
            print(f"[DatasetService] Warning: {path} not found.")
            return

        with open(path, mode="r", encoding="utf-8-sig") as f:
            reader = csv.DictReader(f)
            count = 0
            for row in reader:
                try:
                    self.smoke_records.append({
                        "temp_c": float(row["Temperature[C]"]),
                        "humidity": float(row["Humidity[%]"]),
                        "tvoc_ppb": float(row["TVOC[ppb]"]),
                        "eco2_ppm": float(row["eCO2[ppm]"]),
                        "pm25": float(row["PM2.5"]),
                        "pressure": float(row["Pressure[hPa]"]),
                        "fire_alarm": int(row["Fire Alarm"]),
                    })
                    count += 1
                    if count >= 2000:  # Cap at 2000 records for memory efficiency
                        break
                except Exception:
                    continue

        print(f"[DatasetService] Loaded {len(self.smoke_records)} Smoke IoT records.")

    def get_machine_health(self, asset_id: str = "PUMP-P102") -> Dict[str, Any]:
        if not self.ai4i_records:
            return self._fallback_machine_health(asset_id)

        rec = self.ai4i_records[self.ai4i_index % len(self.ai4i_records)]
        self.ai4i_index += 1

        process_temp_c = round(rec["process_temp_k"] - 273.15, 1)
        air_temp_c = round(rec["air_temp_k"] - 273.15, 1)

        torque_dev = abs(rec["torque_nm"] - 40.0) / 40.0
        wear_factor = min(1.0, rec["tool_wear_min"] / 200.0)
        health_score = max(5.0, round(100.0 - (wear_factor * 50.0) - (torque_dev * 30.0) - (rec["machine_failure"] * 35.0), 1))

        failure_prob = round(75.0 + (rec["machine_failure"] * 20.0), 1) if rec["machine_failure"] else round(max(2.0, (1.0 - health_score / 100.0) * 60.0), 1)
        rul = round(max(0, 14 if rec["machine_failure"] else (200 - rec["tool_wear_min"]) * 0.8), 0)

        failure_mode = None
        if rec["twf"]: failure_mode = "Tool Wear Failure (TWF)"
        elif rec["hdf"]: failure_mode = "Heat Dissipation Failure (HDF)"
        elif rec["pwf"]: failure_mode = "Power Failure (PWF)"
        elif rec["osf"]: failure_mode = "Overstrain Failure (OSF)"
        elif rec["rnf"]: failure_mode = "Random Failure (RNF)"

        return {
            "asset_id": asset_id,
            "name": "Hydrocracker Feed Charge Pump P-102" if asset_id == "PUMP-P102" else f"Machine {rec['product_id']}",
            "product_id": rec["product_id"],
            "category": "PUMP",
            "zone_id": "ZONE-B4",
            "status": "CRITICAL" if health_score < 50 else ("WARNING" if health_score < 75 else "SAFE"),
            "health_score": health_score,
            "air_temp_c": air_temp_c,
            "process_temp_c": process_temp_c,
            "rpm": rec["rpm"],
            "torque_nm": rec["torque_nm"],
            "tool_wear_min": rec["tool_wear_min"],
            "vibration_mm_sec": 8.4 if rec["machine_failure"] else 2.1,
            "operating_hours": 14200,
            "failure_probability": failure_prob,
            "remaining_useful_life": int(rul),
            "failure_mode": failure_mode,
            "machine_failure": bool(rec["machine_failure"]),
        }

    def get_smoke_fire_snapshot(self) -> Dict[str, Any]:
        if not self.smoke_records:
            return {"temp_c": 22.4, "humidity": 55.0, "tvoc_ppb": 12, "eco2_ppm": 400, "pm25": 1.2, "fire_alarm": False, "smoke_level": "NONE"}

        rec = self.smoke_records[self.smoke_index % len(self.smoke_records)]
        self.smoke_index += 1

        is_fire = rec["fire_alarm"] == 1
        smoke_level = "NONE"
        if is_fire:
            smoke_level = "CRITICAL" if rec["tvoc_ppb"] > 5000 else "HIGH"

        return {
            "temp_c": rec["temp_c"],
            "humidity": rec["humidity"],
            "tvoc_ppb": rec["tvoc_ppb"],
            "eco2_ppm": rec["eco2_ppm"],
            "pm25": rec["pm25"],
            "fire_alarm": is_fire,
            "smoke_level": smoke_level
        }

    def _fallback_machine_health(self, asset_id: str) -> Dict[str, Any]:
        return {
            "asset_id": asset_id,
            "name": "Hydrocracker Feed Charge Pump P-102",
            "product_id": "M14865",
            "category": "PUMP",
            "zone_id": "ZONE-B4",
            "status": "CRITICAL",
            "health_score": 42.0,
            "air_temp_c": 25.0,
            "process_temp_c": 35.5,
            "rpm": 1425,
            "torque_nm": 41.9,
            "tool_wear_min": 184,
            "vibration_mm_sec": 8.4,
            "operating_hours": 14200,
            "failure_probability": 88.0,
            "remaining_useful_life": 14,
            "failure_mode": "Overstrain Failure (OSF)",
            "machine_failure": True
        }

dataset_service = DatasetService()
